File: frontend/src/app/domain/practice-rules.ts
import { PracticeAnswerKind, QuestionStats } from './models';

export function nextScore(current: number, result: PracticeAnswerKind): number {
  return result === 'correct' ? current + 1 : current - 1;
}

export function adaptiveWeight(stats: QuestionStats | undefined, sessionMistakeWeight = 0): number {
  const historicalMistakes = (stats?.wrong ?? 0) + (stats?.timeout ?? 0);
  const historicalSuccesses = stats?.correct ?? 0;
  const historyWeight = Math.max(0, historicalMistakes * 2 - historicalSuccesses);
  return 1 + historyWeight + Math.max(0, sessionMistakeWeight);
}

export function pickWeightedIndex<T extends { key: string }>(
  items: T[],
  stats: Record<string, QuestionStats>,
  sessionWeights: Record<string, number>,
): number {
  if (items.length <= 1) return 0;
  const weighted: number[] = [];
  items.forEach((item, index) => {
    const weight = adaptiveWeight(stats[item.key], sessionWeights[item.key] ?? 0);
    for (let count = 0; count < weight; count += 1) weighted.push(index);
  });
  return weighted[Math.floor(Math.random() * weighted.length)] ?? 0;
}

export function normalizePracticeWord(value: string): string {
  return value.trim().toLocaleLowerCase();
}

File: backend/domain/src/main/kotlin/com/example/quiz/PracticeRules.kt
package com.example.quiz

object PracticeScoring {
    fun apply(currentScore: Int, result: PracticeAnswerKind): Int = when (result) {
        PracticeAnswerKind.correct -> currentScore + 1
        PracticeAnswerKind.wrong,
        PracticeAnswerKind.timeout -> currentScore - 1
    }
}

object AdaptiveQuestionWeighting {
    fun weight(stats: QuestionStats = QuestionStats(), sessionMistakeWeight: Int = 0): Int {
        val historicalMistakes = stats.wrong + stats.timeout
        val historicalSuccesses = stats.correct
        val historyWeight = (historicalMistakes * 2 - historicalSuccesses).coerceAtLeast(0)
        return 1 + historyWeight + sessionMistakeWeight.coerceAtLeast(0)
    }

    fun weightedQuestionIds(statsByQuestionId: Map<Long, QuestionStats>, sessionWeights: Map<Long, Int>): List<Long> {
        val ids = (statsByQuestionId.keys + sessionWeights.keys).sorted()
        return ids.flatMap { id ->
            List(weight(statsByQuestionId[id] ?: QuestionStats(), sessionWeights[id] ?: 0)) { id }
        }
    }
}

fun normalizePracticeWord(word: String): String = word.trim().lowercase()
